### ISSUE TITLE
Exception: The user specified as a definer does not exist

### ISSUE THREAD

**Issue by ioweb-gr** [magento/magento2#38422]
### Preconditions and environment

- 2.4.6-p1

### Steps to reproduce

1. Create a database user e.g.  myuser@% 
2. Switch indexers to scheduled mode
3. Switch to production by running all the usual commands setup:upgrade / setup:di-compile etc 
4. Now restrict the database user to only localhost  myuser@localhost
5. Try to create a category and save it. 

You will get an exception until you run setup:upgrade again

### Expected result

User is able to save a category

### Actual result

```
Next Zend_Db_Statement_Exception: SQLSTATE[HY000]: General error: 1449 The user specified as a definer ('db_domains'@'%') does not exist, query was: UPDATE `catalog_category_entity` SET `children_count` = children_count+1 WHERE (entity_id IN('1', '2', '115', '197')) in /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/DB/Statement/Pdo/Mysql.php:109
Stack trace:
#0 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/DB/Statement/Pdo/Mysql.php(91): Magento\Framework\DB\Statement\Pdo\Mysql->tryExecute()
#1 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/zendframework1/library/Zend/Db/Statement.php(313): Magento\Framework\DB\Statement\Pdo\Mysql->_execute()
#2 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php(480): Zend_Db_Statement->execute()
#3 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/zendframework1/library/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract->query()
#4 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/DB/Adapter/Pdo/Mysql.php(564): Zend_Db_Adapter_Pdo_Abstract->query()
#5 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/DB/Adapter/Pdo/Mysql.php(634): Magento\Framework\DB\Adapter\Pdo\Mysql->_query()
#6 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php(635): Magento\Framework\DB\Adapter\Pdo\Mysql->query()
#7 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-catalog/Model/ResourceModel/Category.php(321): Zend_Db_Adapter_Abstract->update()
#8 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-eav/Model/Entity/AbstractEntity.php(1955): Magento\Catalog\Model\ResourceModel\Category->_beforeSave()
#9 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/EntityManager/Observer/BeforeEntitySave.php(36): Magento\Eav\Model\Entity\AbstractEntity->beforeSave()
#10 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Event/Invoker/InvokerDefault.php(88): Magento\Framework\EntityManager\Observer\BeforeEntitySave->execute()
#11 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod()
#12 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Event/Manager.php(65): Magento\Framework\Event\Invoker\InvokerDefault->dispatch()
#13 /var/www/vhosts/domains.gr/httpdocs/generated/code/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager->dispatch()
#14 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/EntityManager/EventManager.php(50): Magento\Framework\Event\Manager\Proxy->dispatch()
#15 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/EntityManager/Operation/Create.php(116): Magento\Framework\EntityManager\EventManager->dispatchEntityEvent()
#16 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/EntityManager/EntityManager.php(106): Magento\Framework\EntityManager\Operation\Create->execute()
#17 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-catalog/Model/ResourceModel/Category.php(1131): Magento\Framework\EntityManager\EntityManager->save()
#18 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Catalog\Model\ResourceModel\Category->save()
#19 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Catalog\Model\ResourceModel\Category\Interceptor->___callParent()
#20 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-catalog-search/Model/Indexer/Fulltext/Plugin/Category.php(44): Magento\Catalog\Model\ResourceModel\Category\Interceptor->Magento\Framework\Interception\{closure}()
#21 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-catalog-search/Model/Indexer/Fulltext/Plugin/Category.php(28): Magento\CatalogSearch\Model\Indexer\Fulltext\Plugin\Category->addCommitCallback()
#22 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(135): Magento\CatalogSearch\Model\Indexer\Fulltext\Plugin\Category->aroundSave()
#23 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Catalog\Model\ResourceModel\Category\Interceptor->Magento\Framework\Interception\{closure}()
#24 /var/www/vhosts/domains.gr/httpdocs/generated/code/Magento/Catalog/Model/ResourceModel/Category/Interceptor.php(41): Magento\Catalog\Model\ResourceModel\Category\Interceptor->___callPlugins()
#25 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Model/AbstractModel.php(658): Magento\Catalog\Model\ResourceModel\Category\Interceptor->save()
#26 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\Model\AbstractModel->save()
#27 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Catalog\Model\Category\Interceptor->___callParent()
#28 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Catalog\Model\Category\Interceptor->Magento\Framework\Interception\{closure}()
#29 /var/www/vhosts/domains.gr/httpdocs/generated/code/Magento/Catalog/Model/Category/Interceptor.php(50): Magento\Catalog\Model\Category\Interceptor->___callPlugins()
#30 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-catalog/Controller/Adminhtml/Category/Save.php(236): Magento\Catalog\Model\Category\Interceptor->save()
#31 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Catalog\Controller\Adminhtml\Category\Save->execute()
#32 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Catalog\Controller\Adminhtml\Category\Save\Interceptor->___callParent()
#33 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Catalog\Controller\Adminhtml\Category\Save\Interceptor->Magento\Framework\Interception\{closure}()
#34 /var/www/vhosts/domains.gr/httpdocs/generated/code/Magento/Catalog/Controller/Adminhtml/Category/Save/Interceptor.php(23): Magento\Catalog\Controller\Adminhtml\Category\Save\Interceptor->___callPlugins()
#35 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/App/Action/Action.php(111): Magento\Catalog\Controller\Adminhtml\Category\Save\Interceptor->execute()
#36 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/module-backend/App/AbstractAction.php(151): Magento\Framework\App\Action\Action->dispatch()
#37 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Backend\App\AbstractAction->dispatch()
#38 /var/www/vhosts/domains.gr/httpdocs/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Catalog\Controller\Adminhtml\Category\Save\Interceptor->___callParent()
```

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by hostep** [comment]
Pretty sure this comes from not stripping the definers from a database dump from some server. If you can figure out where `'db_domains'@'%'` comes from in that `catalog_category_entity` table that should give you a lead.

I use [magerun2](https://github.com/netz98/n98-magerun2/) as a tool to get a database dump from a remote server to import in another db server, it has [build-in support](https://github.com/netz98/n98-magerun2/blob/d0bdf43b31c854ff261d435da3a0a7accee63a5e/src/N98/Magento/Command/Database/DumpCommand.php#L513) to strip away those definer statements as they will usually not match if you use different usernames between different db servers and give errors like you ran into.

Or am I wrong and is this something completely else?

**Comment by ioweb-gr** [comment]
@hostep Not exactly. 

I had my user set up with remote access initially for development purposes. When I finished development, I switched the website to production mode and everything was smooth. Later on I got hit with the exception after I switched the user to not allow remote access. 

So basically during install or setup:upgrade, the definer got set up for my user `@%` and then when the user got updated to only allow `@localhost`, the error popped up.
